src/julia.jl

```diff
@@ -33,4 +33,56 @@ function multireshape(parent::AbstractArray, dimss::Array{Tuple{Vararg{Int}}})
     multireshape(parent, dimss...)
 end
 
+"""
+    pairwise(f::Function, a; kwargs...)
+
+Apply the function `f` to all pairwise combinations of elements from `a`. Set
+keyword `symmetric = true` to only calculate `f(x,y)` when `f(x,y) == f(y,x)`.
+"""
+function pairwise(f::Function, a; symmetric = false)
+    n = length(a)
+    first = f(a[1], a[1])
```

Comment:
`firstval` = f(first(a), first(a))` would handle cases that don't use 1-based indices.  Also, you could pass `firstval` to `pairwise!` (or have an internal `_pairwise!` function that takes an extra argument) to avoid recomputing it.

(This will still be problematic for type-unstable functions.  There is a way to fix this, though...)